jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PullBaseImageStep.java

```diff
@@ -215,139 +218,135 @@ private boolean checkImagePlatform(Image image) {
       RegistryClient registryClient, ProgressEventDispatcher progressEventDispatcher)
       throws IOException, RegistryException, LayerPropertyNotFoundException,
           LayerCountMismatchException, BadContainerConfigurationFormatException {
+    Cache cache = buildContext.getBaseImageLayersCache();
+    EventHandlers eventHandlers = buildContext.getEventHandlers();
+    ImageConfiguration baseImageConfig = buildContext.getBaseImageConfiguration();
+
     ManifestAndDigest<?> manifestAndDigest =
-        registryClient.pullManifest(buildContext.getBaseImageConfiguration().getImageQualifier());
+        registryClient.pullManifest(baseImageConfig.getImageQualifier());
+    eventHandlers.dispatch(
+        LogEvent.lifecycle("Using base image with digest: " + manifestAndDigest.getDigest()));
+
     ManifestTemplate manifestTemplate = manifestAndDigest.getManifest();
+    if (manifestTemplate instanceof V21ManifestTemplate) {
+      V21ManifestTemplate v21Manifest = (V21ManifestTemplate) manifestTemplate;
+      cache.writeMetadata(baseImageConfig.getImage(), v21Manifest);
+      return Collections.singletonList(JsonToImageTranslator.toImage(v21Manifest));
+
+    } else if (manifestTemplate instanceof BuildableManifestTemplate) {
+      // V22ManifestTemplate or OciManifestTemplate
+      BuildableManifestTemplate imageManifest = (BuildableManifestTemplate) manifestTemplate;
+      ContainerConfigurationTemplate containerConfig =
+          pullContainerConfigJson(manifestAndDigest, registryClient, progressEventDispatcher);
+      cache.writeMetadata(baseImageConfig.getImage(), imageManifest, containerConfig);
+      return Collections.singletonList(
+          JsonToImageTranslator.toImage(imageManifest, containerConfig));
+    }
+
+    // TODO: support OciIndexTemplate once AbstractManifestPuller starts to accept it.
+    Verify.verify(manifestTemplate instanceof V22ManifestListTemplate);
```

Comment:
Does this then throw an error for `OciIndexTemplate`? Perhaps we can give the user a better error message (I thought `Verify` is for catching programming/integration issues?)

Reply:
Basically the refactored code doesn't change the current behavior in this regard. The current code would simply break with `ClassCastException` if `manifestTemplate` could ever be of `OciIndexTemplate`, as it can't cast `OciInexTemplate` into `BuildableManifestTemplate` in `jsonManifestToImage()`. That is, it's an impossible condition; we never see `OciIndexTemplate`, as the manifest puller doesn't tell a registry that it can accept an OCI image index. All the registries in practice so far have been returning a Docker manifest list.